environs/config/config.go

```diff
@@ -603,8 +603,8 @@ func Validate(cfg, old *Config) error {
 	}
 
 	if raw, ok := cfg.defined[CloudInitUserDataKey].(string); ok && raw != "" {
-		userDataMap := make(map[string]interface{})
-		if err := yaml.Unmarshal([]byte(raw), &userDataMap); err != nil {
+		userDataMap, err := ensureStringMaps(raw)
+		if err != nil {
 			return errors.Annotate(err, "cloudinit-userdata: must be valid YAML")
```

Comment:
I feel that this may be a slightly confusing error message now since YAML does allow non-string keys, but it is probably fine.

Reply:
reworked the error messages a little, so we only get the "must be valid YAML" if the unmarshal fails